src/igraph_eulerian.c

```diff
@@ -0,0 +1,660 @@
+#include "igraph_adjlist.h"
+#include "igraph_interface.h"
+#include "igraph_components.h"
+#include "igraph_types_internal.h"
+
+/**
+ * \section about_eulerian
+ *
+ * <para>These functions calculate whether an Eulerian path or cycle exists
+ * and if so, can find them.</para>
+ */
+
+
+/* solution adapted from https://www.geeksforgeeks.org/eulerian-path-and-circuit/
+The function returns one of the following values
+has_path is set to 1 if a path exists, 0 otherwise
+has_cycle is set to 1 if a cycle exists, 0 otherwise
+*/
+static int igraph_i_is_eulerian_undirected(const igraph_t *graph, igraph_bool_t *has_path, igraph_bool_t *has_cycle, igraph_integer_t *start_of_path) {
+    igraph_integer_t odd;
+    igraph_vector_t degree, csize, check_for_self_loops, cluster_member;
+    igraph_inclist_t il;
+    long int i, j, vsize;
+    long int cluster_count;
+
+    if (igraph_ecount(graph) == 0 || igraph_vcount(graph) <= 1) {
+        *has_path = 1;
+        *has_cycle = 1;
+        return  IGRAPH_SUCCESS;
+    }
+
+    IGRAPH_VECTOR_INIT_FINALLY(&csize, 0);
+    IGRAPH_VECTOR_INIT_FINALLY(&check_for_self_loops, 0);
+    IGRAPH_VECTOR_INIT_FINALLY(&cluster_member, 0);
+
+    IGRAPH_CHECK(igraph_clusters(graph, &cluster_member, &csize, NULL, IGRAPH_WEAK));
+    cluster_count = 0;
+    vsize = igraph_vector_size(&csize);
+    for (i = 0; i < vsize; i++) {
+        if (VECTOR(csize)[i] > 1) cluster_count++;
+
+        /* extracting all clusters with 1 vertex */
+        if (VECTOR(csize)[i] == 1) {
+            long int vsize2 = igraph_vector_size(&cluster_member);
+            for (j = 0; j < vsize2; j++) {
+                if (VECTOR(cluster_member)[j] == i) {
+                    igraph_vector_push_back(&check_for_self_loops, j);
+                    break;
+                }
+            }
+        }
+
+    }
+
+    if (cluster_count > 1) {
+        *has_path = 0;
+        *has_cycle = 0;
+        igraph_vector_destroy(&csize);
+        igraph_vector_destroy(&check_for_self_loops);
+        igraph_vector_destroy(&cluster_member);
+        IGRAPH_FINALLY_CLEAN(3);
+
+        return IGRAPH_SUCCESS;
+    }
+
+    IGRAPH_CHECK(igraph_inclist_init(graph, &il, IGRAPH_OUT));
+    IGRAPH_FINALLY(igraph_inclist_destroy, &il);
+
+    vsize = igraph_vector_size(&check_for_self_loops);
+    for (i = 0; i < vsize; i++) {
+        igraph_vector_int_t *incedges = igraph_inclist_get(&il, VECTOR(check_for_self_loops)[i]);
+        long int nc = igraph_vector_int_size(incedges);
+        if (nc > 0) {
+            *has_path = 0;
+            *has_cycle = 0;
+            igraph_vector_destroy(&csize);
+            igraph_inclist_destroy(&il);
+            igraph_vector_destroy(&check_for_self_loops);
+            igraph_vector_destroy(&cluster_member);
+            IGRAPH_FINALLY_CLEAN(4);
+
+            return IGRAPH_SUCCESS;
+        }
+    }
+
+    /* assuming that cluster count is not greater than 1 now */
+    /* however, we need to check for self loops still*/
```

Comment:
Were these comments meant to come before the previous loop?

Reply:
moved that